packages/from-hardhat/src/api.ts

```diff
@@ -0,0 +1,142 @@
+import * as Compilation from "./compilation";
+import * as Config from "./config";
+
+import {
+  checkHardhat,
+  askHardhatConsole,
+  askHardhatVersion
+} from "./ask-hardhat";
+import { EnvironmentOptions } from "./options";
+
+/**
+ * Checks for the existence of a Hardhat project configuration and asserts
+ * that the local installed version of Hardhat matches this package's
+ * supported version range.
+ *
+ * @param options to control process environment (e.g. working directory)
+ * @return Promise<void> when expectation holds
+ * @throws NotHardhatError when not in a Hardhat project directory
+ * @throws IncompatibleHardhatError if Hardhat has unsupported version
+ */
+export const expectHardhat = async (
+  options?: EnvironmentOptions
+): Promise<void> => {
+  const isHardhat = await checkHardhat(options);
+
+  if (!isHardhat) {
+    throw new NotHardhatError();
+  }
+
+  const hardhatVersion = await askHardhatVersion(options);
+
+  if (!semver.satisfies(hardhatVersion, supportedHardhatVersionRange)) {
+    throw new IncompatibleHardhatError(hardhatVersion);
+  }
+};
+
+/**
+ * Thrown when no Hardhat project is found
+ */
+export class NotHardhatError extends Error {
+  constructor() {
+    super("Current working directory is not part of a Hardhat project");
+  }
+}
+
+/**
+ * Thrown when Hardhat was detected but with an incompatible version
+ */
+export class IncompatibleHardhatError extends Error {
+  constructor(detectedVersion: string) {
+    super(
+      `Expected Hardhat version compatible with ${supportedHardhatVersionRange}, got: ${detectedVersion}`
+    );
+  }
+}
+
+/**
+ * Constructs a @truffle/config object based on the Hardhat config.
+ *
+ * WARNING: except for fields documented here, the values present on the
+ * returned @truffle/config object MUST be regarded as unsafe to use.
+ *
+ * The returned `config` is defined to contain the following:
+ *
+ *   - `config.networks` with configurations for all Hardhat-configured
+ *     networks, provided:
+ *       - The configured network is not the built-in `hardhat` network
+ *       - The configured network defines a `url` property
+ *
+ *     Note: this function ignores all properties other than `url`,
+ *     including any information that can be used for computing
+ *     cryptographic signatures. THIS FUNCTION DOES NOT READ PRIVATE KEYS.
+ *
+ * Suffice to say:
+ *
+ * THIS FUNCTION'S BEHAVIOR IS EXPERIMENTAL AND SHOULD ONLY BE USED IN
+ * SPECIFICALLY KNOWN-SUPPORTED USE CASES (like reading for configured
+ * network urls)
+ *
+ * @param options to control process environment (e.g. working directory)
+ * @return Promise<TruffleConfig>
+ *
+ * @dev This function shells out to `npx hardhat console` to ask the Hardhat
+ *      runtime environment for a fully populated config object.
+ */
+export const prepareConfig = async (
+  options?: EnvironmentOptions
+): Promise<TruffleConfig> => {
+  const hardhatConfig = (await askHardhatConsole(
+    `hre.config`,
```

Comment:
Worth noting about this approach: this only works because `hre.config` is serializable.

I am so grateful and so jealous that they made this serializable :)

Reply:
Part of the struggle of forging our own path in the early days of the Ethereum ecosystem, I suppose. Also the node ecosystem and what is considered best practice has advanced just a _touch_ since 2015!

Like, I think at the time that the original `truffle-config.js` was designed executable config was the hotness, as it was (and still is) super flexible. Now it's somewhat frowned upon for multi-modal projects like ours due this exact concern.

Reply:
There's trade-offs... like, I also had the thought about how easy it is to read private keys from memory. Our approach clamps that down, since we only store mnemonics inside closures.